Early merges

abakst opened the pull request:
This change introduces a new transform aimed to address issues observed in #478, caused by loops with "early exits", i.e. loops whose branches have postdominators outside the loop. The transform includes two components:

- A pass that identifies loops and any branches to blocks outside the loop;
- A pass that, given a set of branches ("to" and "from" blocks), inserts a single 'router' block that acts as the destination of all of the "to" blocks and then branches to the correct "from" block. That is, given pairs of blocks (bin_i, bout_j), conceptually the transformation:
  1. Invents a register `r`, and in each block `bin_i` sets `r := inj_j()` where `inj_j()` is the `j-th` injection in a sum type
  2. Invents a block `route` a modifies each `bin_i` to jump to `route`. `route` then switches on the value of `r`: if `r == inj_j()` then control flow jumps to `bout_j` .

This PR exposes this transform as an option in `crux-llvm`.

abakst commented:
I've marked this as WIP as I still need to properly integrate some tests, but the implementation should be ready for review

travitch commented:
I don't know if this is entirely relevant to you, but I saw that you have some loop finding code in here.  There is another implementation of loop finding in crucible already https://github.com/GaloisInc/crucible/blob/master/crucible/src/Lang/Crucible/Analysis/Fixpoint/Components.hs. It might not be as useful to you if you need more granularity on the sub-components of a loop.

robdockins commented on crucible/src/Lang/Crucible/CFG/Transforms.hs:
```diff
@@ -0,0 +1,730 @@
+      | Just blk' <- Map.lookup blkId bmap'
+      = traverseStep f bmap' genv' penv' seen' blk'
+      | otherwise
+      = undefined
```
Don't like to see `undefined`.  Let's `panic` if we beleive this cannot happen

robdockins commented on crucible/src/Lang/Crucible/CFG/Transforms.hs:
```diff
@@ -0,0 +1,730 @@
+newtype PartialValue s tp = PartialValue { getPartial :: Reg s (MaybeType tp) }
+type ValueToPartialMap s  = MapF.MapF (Value s) (PartialValue s)
+  
+traverseCFG :: ( Monad m, TraverseExt ext )
```
This seems like it might be generally reusable... maybe worth pulling into a more generic module?

abakst replied:
Does it make sense to move it into `Reg` (which seems to be the main module for mutable-register based CFGs)?

robdockins replied:
Yeah, it might make sense there.

robdockins reviewed (changes requested):
This is a hard algorithm to review, but I'll say that I don't see anything obviously wrong, so that's a good start :-)

I'm probably not going to be happy merging this until I see some kind of unit tests that show at least some simple examples of this algorithm working and producing the expected results.  I'll also want to see a lot more documentation on the internal functions, as a lot of what's going on in here is pretty subtle.

Finally, the git history for this PR has a lot of messy intermediate commits.  I'd like to see it rebased into something more organized.

abakst commented on crucible/src/Lang/Crucible/CFG/Transforms.hs:
```diff
@@ -0,0 +1,730 @@
+   li1 <= li2 = isNested li1 li2
+
+-- | Detect all loops in a cfg.
+-- The assumption is that two backedges in a cfg will have distinct destination blocks.
```
You're right, I'm not sure that it should be in general. While I'm not convinced it will matter in practice if we discard that case (i.e., don't apply the transformation), it might be worth double checking if this assumption is critical. For now, I think this assumption ought to be checked.

abakst commented on crucible/src/Lang/Crucible/CFG/Transforms.hs:
```diff
@@ -0,0 +1,730 @@
+      _ -> return mempty
+
+    lowerBlockOutput = case blockTerm blk of
+      Posd p (Output ll a) ->
```
I'll write a proper explanation. However, if I understand correctly, these are the only term-statements that output a value (i.e. an `Atom`) to the destination block. This Atom is the (potential) target of the "lowering".

abakst replied:
Actually on further reflection, there should be a few others... thanks :)

abakst commented on crucible/src/Lang/Crucible/CFG/Transforms.hs:
```diff
@@ -0,0 +1,730 @@
+-- Create a single block that merges all paths before
+-- continuing on to the respective j's
+--
+-- Create a unique block F and multiple blocks  i' j' such
```
Yes, that's right. It feels a bit cumbersome -- if there's a better way to do it I'm interested.

robdockins replied:
Humm... not offhand.

abakst commented on crucible/src/Lang/Crucible/CFG/Transforms.hs:
```diff
@@ -0,0 +1,730 @@
+
+------------------------------------------------------------------------
+-- Fixup Pass
+              -- HEREHEREHERE need to account for lambda block lowered atoms
```
Eek, that one was leftover, thanks.

abakst commented:
I believe I've addressed your comments @robdockins. I also added some tests and tweaked how the test suite works: basically the test suite can be organized into subdirectories and the tests will be grouped accordingly. This seemed like a convenient way to organize any potential future test suites, but I'm not especially committed to this structure.

robdockins commented on crucible/src/Lang/Crucible/CFG/EarlyMergeLoops.hs:
```diff
@@ -0,0 +1,839 @@
+type CFGEdge s = (BlockID s, BlockID s)
+
+-- | Detect all loops in a cfg.
+-- The assumption is that two backedges in a cfg will have distinct destination blocks.
```
If this assumption is violated, then we return the empty list of loop info and do nothing, is that correct?  Let's state that in this comment.

robdockins commented on crucible/src/Lang/Crucible/CFG/EarlyMergeLoops.hs:
```diff
@@ -0,0 +1,839 @@
+     let cfg' = cfg { cfgBlocks = blks' }
+     return cfg'
+  where
+    root = fromMaybe (error "Unknown root") $ findBlock cfg rootLabel
```
Let's use `panic` here instead.

robdockins commented on crux-llvm/src/Crux/LLVM/Config.hs:
```diff
@@ -107,6 +108,9 @@ llvmCruxConfig =
                            "Optimization level to request from `clang`"
 
+         loopMerge <- Crux.section "opt-loop-merge" Crux.yesOrNoSpec False
+                        "Insert merge blocks in loops with early exits"
```
Maybe add a short sentence here about why you'd want to do this.  "may speed up simulation", etc.

robdockins commented:
This is in much better shape than before, thanks for working on this @abakst!

I have some minor comments above, and a slightly more serious comment here.  I'm still a little concerned about the algorithmic structure of `earlyMergeLoops`.  I think I understand why we need to reanalyze the structure of the loops after every iteration, but are we pretty sure we can't accidentally get into an infinite loop?  We're adding new blocks, but I guess they can never be the target of a back-edge, and that's enough to ensure that the keeping track of the seen blocks means we eventually terminate?

Second, I wonder about running `lowerUndefPass` after each iteration inside `earlyMergeLoop`... isn't it possible that the same register could be lowered multiple times, resulting in an unnecessary tower of `Maybe` types and projections?  Would it work to run `lowerUndefPass` just once after all the loops have been defunctionalized?

robdockins reviewed (changes requested):
Minor changes and some questions; see above.  Otherwise, I think this is ready to go.

abakst commented:
@robdockins I believe I've addressed your concerns at this point, with one additional change -- I've checked in the optimization tests as LLVM bitcode (disassembled) so that the tests remain stable across clang versions.

It's worth recording here that for some of the tests, clang version 10 produces CFGs whose performance with and/or without the optimization is quite different from the CFG produced by version 9. It might be worth investigating this more in the future to tune how we decide which loops should be transformed.

I think the remaining CI failures (on macOS only) are unrelated to this change.

robdockins commented on crucible/src/Lang/Crucible/CFG/EarlyMergeLoops.hs:
```diff
@@ -541,22 +542,51 @@ earlyMergeLoops :: ( TraverseExt ext, Monad m, Show (CFG ext s init ret) )
+       let seen'    = (Set.insert (liHeader l) seen)
+       let nextStep = candidates seen' ls'
+       -- The termination transition invariant is that 
+       when (length thisStep <= length nextStep) $
```
Nitpick, this should be `<` for termination.

robdockins reviewed (approved):
Let's keep the original `.c` files that produced the `.ll` around instead of deleting them (so we can update them in the future if necessary), but otherwise this looks good!
